4-Beyond-Classical-Search/c_hillClimbing.js

```diff
@@ -1,145 +1,192 @@
+        .attr('height', d => this.yScale(d.objective))
+        .attr('width', this.blockWidth)
+        .attr('x', (d) => this.xScale(d.state))
+        .attr('y', (d) => this.h - this.yScale(d.objective))
+        .style('opacity', (d) => (d.visited) ? 1 : 0)
+        .style('fill', 'hsl(217,61.2%,53.5%)')
+        .style('border', '1px solid');
+    }
+
+    visit(state) {
+      this.hillData[state].visited = true;
+      this.svgRects.filter((d) => d.state == state)
+        .transition()
+        .duration(100)
+        .style('opacity', 1);
+    }
+
+    showAll() {
+      this.svgRects.transition()
+        .duration(100)
+        .style('opacity', (d) => {
+          return (d.visited) ? 1 : 0.6;
+        })
+        .style('fill', (d) => {
+          if (d.maxima) {
+            if (d.visited) {
+              return 'hsl(110, 100%, 38%)';
+            } else {
+              return 'hsl(102, 100%, 56%)';
+            }
+          } else {
+            return 'hsl(217,61.2%,53.5%)';
+          }
+        })
+    }
+  }
+
+	//Class to draw the robot
+  class HillClimberDiagram {
+    constructor(h, w, svg, hill, hillDiagram, hillClimber) {
+      this.h = h;
+      this.w = w;
+      this.hill = hill;
+      this.hillDiagram = hillDiagram;
+      this.hillClimber = hillClimber;
+      this.states = this.hill.getStates();
+      this.svg = svg;
+      this.xScale = this.hillDiagram.xScale;
+      this.yScale = this.hillDiagram.yScale;
+      this.blockWidth = this.hillDiagram.blockWidth;
+      this.renderHillClimber();
+    }
+    renderHillClimber() {
+      let robotLocation = this.hillClimber.getCurrentState();
+      let robotStateValue = this.states[robotLocation];
+      this.robot = this.svg.append('g')
+        .attr('class', 'robot')
+        .append('rect')
+        .attr('height', '10px')
+        .attr('width', this.blockWidth)
+        .attr('x', this.xScale(robotLocation))
+        .attr('y', this.h - this.yScale(robotStateValue) - 10)
+        .style('fill', 'red');
+      this.hillDiagram.visit(robotLocation);
+
+    }
+    move(state) {
+      let robotLocation = state;
+      let robotStateValue = this.states[state];
+      this.robot.transition()
+        .duration(100)
+        .attr('x', this.xScale(robotLocation))
+        .attr('y', this.h - this.yScale(robotStateValue) - 10);
+    }
+  }
+
+	//Wrapper class for the entire diagram
+  class HillClimbingDiagram {
+    constructor(selector, h, w) {
+      this.h = h;
+      this.w = w;
+      this.svg = d3.select(selector).html("")
+        .append('svg')
+        .attr('height', this.h)
+        .attr('width', this.w)
+        .attr('border', 2);
+
+      this.hill = new Hill();
+      this.hillClimber = new HillClimber(this.hill);
+      this.hillDiagram = new HillDiagram(this.svg, this.hill, this.h, this.w);
+      this.hillClimberDiagram = new HillClimberDiagram(this.h, this.w, this.svg, this.hill, this.hillDiagram, this.hillClimber);
+      this.bindClicks();
```

Comment:
Minor suggestion: the second constructor takes svg, hill, h, w, and the third constructor takes h, w, svg, hill; it'd look a little nicer if they used a consistent order